### PR TITLE
[clang] replaced the usage of `asctime` with `std::put_time`

### LINKED ISSUE #98724: clang/lib/Lex/PPMacroExpansion.cpp: 2 * calls to obsolete function asctime
Static analyser cppcheck says:

1.

trunk/clang/lib/Lex/PPMacroExpansion.cpp:1728:16: style: Obsolete function 'asctime' called. It is recommended to use 'strftime' instead. [asctimeCalled]

Source code is

     Result = asctime(TM);

2.

trunk/clang/lib/Lex/PPMacroExpansion.cpp:1738:18: style: Obsolete function 'asctime' called. It is recommended to use 'strftime' instead. [asctimeCalled]

Source code is

      Result = asctime(TM);

But "man asctime" says:

    Marked obsolete in POSIX.1-2008 (recommending strftime(3)).

2008 is 16 years ago.

### PR DESCRIPTION
**Pull request by yichi170**
In `clang/lib/Lex/PPMacroExpansion.cpp`, replaced the usage of the obsolete `asctime` function with `std::put_time` for generating timestamp strings.

Fixes: https://github.com/llvm/llvm-project/issues/98724

### CODE REVIEW

File: clang/lib/Lex/PPMacroExpansion.cpp
Comment (shafik) [inline clang/lib/Lex/PPMacroExpansion.cpp:1729]: I believe we need to check the result of `strftime` is not `0` which would indicate that it ran out of space and we need to recover.

Comment (yichi170) [reply clang/lib/Lex/PPMacroExpansion.cpp:1729]: I see. Do you mean if the result of `strftime` is zero, `Result` have to be set as "???...?"?

Comment (AaronBallman) [reply clang/lib/Lex/PPMacroExpansion.cpp:1729]: I think there's three problems here:

1) We need to look at the return value from `strftime()` and if it's `0`, fallback to the failure mode.
2) Different locales may require a different buffer size.
3) `%c` isn't specified to return the same thing as `asctime()` though it frequently will.

So I think `TimeString` should probably have a hard-coded buffer size that's a bit bigger than we expect is necessary, like `char TimeString[64];`, and the format string to `strftime()` should be `%a %b %d %T %Y`, and we should check the return value of the call to `strftime()` and fall back to `?? .. ?` if the value is 0.

**Comment by cor3ntin** [comment]
`asctime` is deprecated. But `asctime_s` is not.
See https://en.cppreference.com/w/c/chrono/asctime

using `strftime` is going to be challenging because of locale (ie we would have to temporarily switch to a C locale - both because we want a locale independent behavior, and because for an arbitrary locale the buffer might need to be bigger than for the C locale.

(an alternative would be to use the C++ interface https://compiler-explorer.com/z/WEqTW341b )

Either way, we need to make sure sure the produced string are identical, I'm sure people rely on that.

File: clang/lib/Lex/PPMacroExpansion.cpp
Comment (cor3ntin) [inline clang/lib/Lex/PPMacroExpansion.cpp:1728]: As an aside, I find it weird that we use gmtime in one case and localtime in the other (see L1739)

File: clang/lib/Lex/PPMacroExpansion.cpp
Comment (yichi170) [inline clang/lib/Lex/PPMacroExpansion.cpp:1729]: @AaronBallman I adapted your advice and here I use `%e` instead of `%d` since `asctime` returns the space-padded Day of the month.

File: clang/lib/Lex/PPMacroExpansion.cpp
Comment (AaronBallman) [reply clang/lib/Lex/PPMacroExpansion.cpp:1728]: It's consistent with the behavior in `ComputeDATE_TIME`, perhaps because `SourceDateEpoch` is the number of seconds since the epoch and it's easier to reason about that from UTC as a base than localtime (e.g. for build reproducibility)

File: clang/lib/Lex/PPMacroExpansion.cpp
Comment (AaronBallman) [inline clang/lib/Lex/PPMacroExpansion.cpp:1729]: Hmm, so the last remaining problem is that of locale. I didn't realize that `asctime()` is locale independent, which means that switching to `strftime()` will change the output for some users. :-(

We could use `setlocale()` before the call to `strftime()` and then reset the locale after the call, but another approach would be to refactor `ComputeDATE_TIME` which does all the formatting manually and in a locale-independent way. Given what a pain locales are in practice, I sort of think the refactoring is actually a better approach. WDYT @cor3ntin?

Comment (cor3ntin) [reply clang/lib/Lex/PPMacroExpansion.cpp:1729]: This is why i suggested the C++ approach here https://github.com/llvm/llvm-project/pull/99075#issuecomment-2233382915 

(wouldn't it be nice if the C standard provided overloads of everything that takes explicit local parameters to avoid having to do these sort of dances!)

Comment (yichi170) [reply clang/lib/Lex/PPMacroExpansion.cpp:1729]: I found that using `%b` in `put_time` is also locale-dependent. Is it expected?

Comment (AaronBallman) [reply clang/lib/Lex/PPMacroExpansion.cpp:1729]: SHOOT, yes, that is expected, same for `%a`.

> This is why i suggested the C++ approach here https://github.com/llvm/llvm-project/pull/99075#issuecomment-2233382915

We can't use that approach because we need to compile in C++17 mode: https://compiler-explorer.com/z/xosc68WMn

Okay, so how's this for horrible:
```
bool Failed = false;
if (const char *OldLocale = ::setlocale(LC_TIME, "C")) {
  if (strftime(...)) {
  } else {
    Failed = true;
  }
  ::setlocale(LC_TIME, OldLocale);
} else {
  Failed = true;
}

if (Failed) {
  // Emit question marks instead
}
```

Comment (cor3ntin) [reply clang/lib/Lex/PPMacroExpansion.cpp:1729]: @AaronBallman the `std::print` thing was illistrative, this works in C++17 (well, 11)

```cpp
    std::stringstream buf;
    // this is the line that ensure we are not affected by locales 
    buf.imbue(std::locale("C"));
    std::time_t t = std::time(nullptr);
    std::tm tm = *std::localtime(&t);
    buf << std::put_time(&tm, "%c");
    std::string str = buf.str(); // do whatever with that string
```

Comment (AaronBallman) [reply clang/lib/Lex/PPMacroExpansion.cpp:1729]: Ah, okay, thanks! Hmm, if we have to mess with locales either way, this seems reasonable to me.

Comment (MaskRay) [reply clang/lib/Lex/PPMacroExpansion.cpp:1729]: `asctime_r` might be the best replacement, but it seems unavailable on MSVC.
MSVC has `asctime_s`.

**Comment by yichi170** [comment]
I implemented with `std::put_time` instead, and `setstate` for further checking whether `std::put_time` is successful or not.

File: clang/lib/Lex/PPMacroExpansion.cpp
Comment (cor3ntin) [inline clang/lib/Lex/PPMacroExpansion.cpp:1729]: That is not needed afaik (the bad bit would be set automatically on failure)

Comment (yichi170) [reply clang/lib/Lex/PPMacroExpansion.cpp]: Fixed it. Thanks for the advice.

**Review by AaronBallman — APPROVED** [review]
Thank you, the changes LGTM assuming precommit CI doesn't discover anything. @cor3ntin are you happy with the changes as well?

**Comment by yichi170** [comment]
I don't have write access, so if the PR is ready, please help me land it!
If there is anything that I need to do, please let me know!

**Comment by cor3ntin** [comment]
LGTM too!
Thanks for the fix :)

**Comment by rorth** [comment]
This seems to have broken the [Solaris/amd64](https://lab.llvm.org/staging/#/builders/120/builds/187) and [Solaris/sparcv9](https://lab.llvm.org/buildbot/#/builders/13/builds/1030) buildbots.

**Comment by rorth** [comment]
Confirmed: reverting the change locally restores the builds, although I don't yet see why.

**Comment by AaronBallman** [comment]
> Confirmed: reverting the change locally restores the builds, although I don't yet see why.

The fact that there's a missing symbol suggests the STL on the machine is not conforming... the symbol that's missing  (`_ZNKSt8time_putIcSt19ostreambuf_iteratorIcSt11char_traitsIcEEE3putES3_RSt8ios_basecPKSt2tmPKcSB_`) demangles to: `std::time_put<char, std::ostreambuf_iterator<char, std::char_traits<char> > >::put(std::ostreambuf_iterator<char, std::char_traits<char> >, std::ios_base&, char, std::tm const*, char const*, char const*) const` which has been around since C++98.

**Comment by AaronBallman** [comment]
Aha!

https://bugs.llvm.org/show_bug.cgi?id=33767

We have a mixture of both `std::tm` and `::tm` in here, try switching to using `::tm` and see if that helps.

**Comment by rorth** [comment]
> Aha!
> 
> https://bugs.llvm.org/show_bug.cgi?id=33767

Ah, old sins (from the GCC side no less) keep haunting me.

> We have a mixture of both `std::tm` and `::tm` in here, try switching to using `::tm` and see if that helps.

Unfortunately not: I always get the undefined reference to the `std::tm const *` version, no matter what I tried.

**Comment by AaronBallman** [comment]
> > We have a mixture of both `std::tm` and `::tm` in here, try switching to using `::tm` and see if that helps.
> 
> Unfortunately not: I always get the undefined reference to the `std::tm const *` version, no matter what I tried.

Drat!

If you dump the symbols from the STL shared library on the system, is there one for `time_put` at all? If so, what is the mangled symbol it exports?

(As a perhaps terrible idea, could we use the `alias` attribute on a redeclaration to try to force to link against the correct mangling for Solaris only?)

**Comment by rorth** [comment]
> If you dump the symbols from the STL shared library on the system, is there one for `time_put` at all? If so, what is the mangled symbol it exports?

While the source requires
```
_ZNKSt8time_putIcSt19ostreambuf_iteratorIcSt11char_traitsIcEEE3putES3_RSt8ios_basecPKSt2tmPKcSB_

std::time_put<char, std::ostreambuf_iterator<char, std::char_traits<char> > >::put(std::ostreambuf_iterator<char, std::char_traits<char> >, std::ios_base&, char, std::tm const*, char const*, char const*) const
```
`libstdc++.so.6` only provides
```
_ZNKSt8time_putIcSt19ostreambuf_iteratorIcSt11char_traitsIcEEE3putES3_RSt8ios_basecPK2tmPKcSB_

std::time_put<char, std::ostreambuf_iterator<char, std::char_traits<char> > >::put(std::ostreambuf_iterator<char, std::char_traits<char> >, std::ios_base&, char, tm const*, char const*, char const*) const
```
i.e. `tm const *` instead of `std::tm const *`.
> 
> (As a perhaps terrible idea, could we use the `alias` attribute on a redeclaration to try to force to link against the correct mangling for Solaris only?)

A hack along the lines of
```
asm("_ZNKSt8time_putIcSt19ostreambuf_iteratorIcSt11char_traitsIcEEE3putES3_RSt8ios_basecPKSt2tmPKcSB_ = _ZNKSt8time_putIcSt19ostreambuf_iteratorIcSt11char_traitsIcEEE3putES3_RSt8ios_basecPK2tmPKcSB_");
```
at least allows `clang` to link.

**Comment by AaronBallman** [comment]
One last thing to try -- what happens if you replace `::tm` and `std::tm` with `struct tm`? Does that happen to get you the correct symbol (hoping the elaborated type name makes a difference)?

**Comment by AaronBallman** [comment]
Blech, thank you for trying! I guess that unless there's some way to hide this ugliness in the cmake scripts or someone else has better ideas to try, the `asm` aliasing (on Solaris only) may be our best path forward.

**Comment by rorth** [comment]
At least in the short term.  ISTM that `clang++` may need to do something along the lines of Bug 33767 (resp. the underlying GCC bug/patch), otherwise users are in for a nasty surprise.

**Comment by AaronBallman** [comment]
> At least in the short term. ISTM that `clang++` may need to do something along the lines of Bug 33767 (resp. the underlying GCC bug/patch), otherwise users are in for a nasty surprise.

Agreed; long-term we need a better solution. Hopefully someone maintaining Solaris support can step in and help there, because this is quite surprising behavior.

**Comment by yichi170** [comment]
Thanks for identifying the problem. @AaronBallman @rorth

**Comment by rorth** [comment]
> > At least in the short term. ISTM that `clang++` may need to do something along the lines of Bug 33767 (resp. the underlying GCC bug/patch), otherwise users are in for a nasty surprise.
> 
> Agreed; long-term we need a better solution. Hopefully someone maintaining Solaris support can step in and help there, because this is quite surprising behavior.

The cleaned-up hack is in the final rounds of testing.  Wlll create a PR once that's done.

There's no real Solaris maintainer AFAIK: while I do run the buildbots and try to keep the port in shape as time allows, I can only do so much.  The Solaris GCC maintainership takes enough of my time and splitting it would only be to the detriment of both projects.

TBH, I wouldn't even know where to look: I've mostly dealt with `compiler-rt` and driver issues so far.

**Pull request merged**
